juju/client/connection.py

```diff
@@ -649,14 +649,12 @@ def _build_facades(self, facades):
             # the following attempts to get the best facade version for the
             # client. The client knows about the best facade versions it speaks,
             # so in order to be compatible forwards and backwards we speak a
-            # common facade version.
+            # common facade versions.
             if name in client_facades:
-                self.facades[name] = client_facades[name]['version']
-            else:
-                # if a facade is passed back that we don't know about, just use
-                # the latest version, so we don't break against new controller
-                # versions.
-                self.facades[name] = facade['versions'][-1]
+                known = client_facades[name]['versions']
+                discovered = facade['versions']
+                version = max(set(known).intersection(set(discovered)))
```

Comment:
we should probably handle:
```
>>> x = [1, 2]
>>> y = [3, 4]
>>> max(set(x).intersection(y))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: max() arg is an empty sequence
```

```
try:
  version = ...
except ValueError:
  # facade does not have a common version, so it isn't a facade we can support
else:
  self.facades[name] = version
```

Reply:
Good shout, didn't think about that one

Reply:
@jameinel i've updated this in the subsequent commit https://github.com/juju/python-libjuju/pull/321/commits/7582e5208263c274b615ceceb642f9e769a24011